### ISSUE TITLE
Just an idea for discussion

### ISSUE THREAD

**Issue by 072jiajia** [openai/CLIP#296]
Hi! Thanks for sharing the great work!
I have some questions that I want to discuss with the authors and everyone interested in this work.
In Brief, do you think using the following loss function can make the training process more stable? Or can it make the model utilize the joint multi-modal embedding spaces more efficiently?

Numpy-like pseudocode
```python=
# image_encoder - ResNet or Vision Transformer
# text_encoder - CBOW or Text Transformer
# I[n, h, w, c] - minibatch of aligned images
# T[n, l] - minibatch of aligned texts
# W_i[d_i, d_e] - learned proj of image to embed
# W_t[d_t, d_e] - learned proj of text to embed
# W_std_i[d_i, ] - learned proj of image to standard deviation in logarithmic form
# W_std_t[d_t, ] - learned proj of text to standard deviation in logarithmic form

# extract feature representations of each modality
I_f = image_encoder(I)  # [n, d_i]
T_f = text_encoder(T)  # [n, d_t]
# joint multimodal embedding [n, d_e]
I_e = np.dot(I_f, W_i)
T_e = np.dot(T_f, W_t)
# logarithmic form standard deviation in joint multimodal embedding space
I_ln_std = np.expand_dims(np.dot(I_f, W_std_i), axis=1) # [n, 1]
T_ln_std = np.expand_dims(np.dot(T_f, W_std_t), axis=0) # [1, n]
# variance in joint multimodal embedding space
I_var = np.exp(I_ln_std)**2 # [n, 1]
T_var = np.exp(T_ln_std)**2 # [1, n]
# pairwise distances [n, n]
distances_sq = np.sum((np.expand_dims(I_f, axis=1) - np.expand_dims(T_f, dim=0)) ** 2, axis=2)
# symmetric loss function
labels = np.arange(n)
loss_i = cross_entropy_loss(-I_ln_std * distances / I_var, label, axis=0)
loss_t = cross_entropy_loss(-T_ln_std * distances / T_var, label, axis=1)
loss = (loss_i + loss_t)/2

```

PyTorch-like pseudocode
```python=
def my_loss(I_f, I_ln_std, T_f, T_ln_std):
    '''
        I_f: [n, d_e]
        T_f: [n, d_e]
        I_ln_std: [n, ]
        T_ln_std: [n, ]
    '''
    n = I_f.size(0)
    ce_loss = nn.CrossEntropyLoss()

    # I_f: [n, 1, d_e]
    I_f = I_f.unsqueeze(1)
    # T_f: [1, n, d_e]
    T_f = T_f.unsqueeze(0)
    # [n, n]
    distances = torch.sum((I_f - T_f) ** 2, dim=2)

    # [n, ]
    I_var = I_ln_std.exp() ** 2
    T_var = T_ln_std.exp() ** 2

    # [1, n]
    T_ln_std = T_ln_std.unsqueeze(0)
    T_var = T_var.unsqueeze(0)
    I_ln_std = I_ln_std.unsqueeze(0)
    I_var = I_var.unsqueeze(0)

    label = torch.arange(n, device=I_f.device)
    loss_t = ce_loss(-T_ln_std * distances / T_var, label)
    loss_i = ce_loss(-I_ln_std * distances.t() / I_var, label)

    return (loss_t + loss_i) / 2


I_f, I_ln_std = image_encoder(image)
T_f, T_ln_std = text_encoder(text)

loss = my_loss(I_f, I_ln_std, T_f, T_ln_std)
```

**Comment by 072jiajia** [comment]
Considering the following two sentences, "A dog in the garden" and "A shiba dog in the garden", it is clear that the second sentence has more conditions. We may say that the first sentence's variance should be larger, and the second sentence's variance should be smaller.
To be more precise, the image with more specific information should have lower variance for output.

**Comment by 072jiajia** [comment]
Consider the following annotations:
* $img_i: \text{The i-th image}, \\
text_i: \text{The i-th text}, \\
I: \text{The set of image's features}, \\
T: \text{The set of text's features}, \\
I_i: \text{The feature of the i-th image}, \\
T_i: \text{The feature of the i-th text}, \\
\sigma_{img_i}: \text{The standard deviation of the i-th image}, \\
\sigma_{t_i}: \text{The standard deviation of the i-th text}, \\
Batch: \text{A batch of data}$

First, assuming that the feature's distribution is a guassion distribution. If we have a text $T$, then the probability of the corresponding image is $I$ can be written as
$P(img | text, Batch) = P(img | text) = f(I - T, \sigma_{t}) = 
\frac{1}{\sigma_{t} \sqrt{2 \pi}} exp (- \frac{(I - T)^2}{2  \sigma_{t}^2 })$


By Bayes' theorem, 
$P(text_a|img_b, Batch) = \frac
{P(img_b|text_a, Batch) P(text_a | Batch)}
{\sum\limits_i P(img_b|text_i, Batch) P(text_i | Batch)}$

As $P(text_i|Batch)$ are the same, we have
$P(text_a|img_b, Batch) = \frac
{P(img_b|text_a)}
{\sum\limits_i  P(img_b|text_i)} = \frac
{\frac{1}{\sigma_{t_a}} exp (- \frac{(I_b - T_a)^2}{2  \sigma_{t_a}^2 })}
{\sum\limits_i  \frac{1}{\sigma_{t_i}} exp (- \frac{(I_i - T_i)^2}{2  \sigma_{t_i}^2 })} = SoftMax(\{-ln(\sigma_{t_i})\frac{(I_b - T_i)^2}{2  \sigma_{t_i}^2 }\}_{i=1}^{N})_a$


Therefore, we can formulate the cross entropy loss as below
$CrossEntropyLoss(\{-ln(\sigma_{t_i})\frac{(I_b - T_i)^2}{2  \sigma_{t_i}^2 }\}_{i=1}^{N})$

Symmetrically, we can have 
$CrossEntropyLoss(\{-ln(\sigma_{img_i})\frac{(T_a - I_i)^2}{2  \sigma_{img_i}^2 }\}_{i=1}^{N})$



We can observe that the loss used in the original paper can also be explained in this form.
Consider the computation of cosine similarity
$|I_j|_2 = 1, |T_i|_2 = 1, I_j T_i = cos \theta_{ji}$

We can have the following obsevation
$CrossEntropyLoss(\{-ln(\sigma_{t_i})\frac{(I_j - T_i)^2}{2  \sigma_{t_i}^2 }\}_{i=1}^{N}) \\
= CrossEntropyLoss(\{-\frac{ln(\sigma_{t_i})}{2  \sigma_{t_i}^2}(|I_j|_2^2 + |T_i|_2^2 - 2  I_j T_i)\}_{i=1}^{N}) \\
= CrossEntropyLoss(\{-C_1(2 - 2 cos \theta_{ji}\}_{i=1}^{N}) \text{ \\\\ Do some substitution}\\
= CrossEntropyLoss(\{-2C_1 + 2C_1 cos \theta_{ji}\}_{i=1}^{N}) \text{ \\\\ the -2C1 can be removed}\\
= CrossEntropyLoss(\{C_2 cos \theta_{ji}\}_{i=1}^{N})$
Where $C_2$ is the np.exp(t) learned in the paper

**Comment by 072jiajia** [comment]
Sorry I'm searching for some way to write equations here
The remaining ideas are in here
https://hackmd.io/@jS3Mpow5SSOX1M0cbyflTw/Hy9w75-Qs

**Comment by 072jiajia** [comment]
In the view of probability
![a](https://user-images.githubusercontent.com/63368701/197477077-5679c0a5-06cb-459d-a691-69087d1d96e5.png)
![b](https://user-images.githubusercontent.com/63368701/197477149-a9c45f00-81a3-48dc-ab27-8bc2c28f523b.png)
![c](https://user-images.githubusercontent.com/63368701/197477159-da03440e-6253-4584-88f5-f276dfc3f7a2.png)

**Comment by 072jiajia** [comment]
Thank you for reading util here. Please tell me if my thought has any mistakes. Since I don't have that many GPUs and corresponging computer equipments to do the experiment, if you also think the idea make sense, please try the experiment. I'm looking forward to knowing that whether this method actually works.
